crates/viewer/re_space_view_graph/src/layout/provider.rs

```diff
@@ -93,116 +107,175 @@ impl ForceLayoutProvider {
-    let mut tmin = f32::NEG_INFINITY;
-    let mut tmax = f32::INFINITY;
-
-    for i in 0..2 {
-        let inv_d = 1.0 / -direction[i];
-        let mut t0 = (rect.min[i] - rect.center()[i]) * inv_d;
-        let mut t1 = (rect.max[i] - rect.center()[i]) * inv_d;
-
-        if inv_d < 0.0 {
-            std::mem::swap(&mut t0, &mut t1);
-        }
-
-        tmin = tmin.max(t0);
-        tmax = tmax.min(t1);
+    PathGeometry::Line {
+        source: source_point,
+        target: target_point,
     }
-
-    let t = tmax.min(tmin); // Pick the first intersection
-    rect.center() + t * -direction
 }
 
-#[cfg(test)]
-mod test {
-    use super::*;
-    use egui::pos2;
-
-    #[test]
-    fn test_ray_intersection() {
-        let rect = Rect::from_min_max(pos2(1.0, 1.0), pos2(3.0, 3.0));
-
-        assert_eq!(
-            intersects_ray_from_center(rect, Vec2::RIGHT),
-            pos2(3.0, 2.0),
-            "rightward ray"
-        );
-
-        assert_eq!(
-            intersects_ray_from_center(rect, Vec2::UP),
-            pos2(2.0, 1.0),
-            "upward ray"
-        );
-
-        assert_eq!(
-            intersects_ray_from_center(rect, Vec2::LEFT),
-            pos2(1.0, 2.0),
-            "leftward ray"
-        );
-
-        assert_eq!(
-            intersects_ray_from_center(rect, Vec2::DOWN),
-            pos2(2.0, 3.0),
-            "downward ray"
-        );
-
-        assert_eq!(
-            intersects_ray_from_center(rect, (Vec2::LEFT + Vec2::DOWN).normalized()),
-            pos2(1.0, 3.0),
-            "bottom-left corner ray"
-        );
-
-        assert_eq!(
-            intersects_ray_from_center(rect, (Vec2::LEFT + Vec2::UP).normalized()),
-            pos2(1.0, 1.0),
-            "top-left corner ray"
-        );
-
-        assert_eq!(
-            intersects_ray_from_center(rect, (Vec2::RIGHT + Vec2::DOWN).normalized()),
-            pos2(3.0, 3.0),
-            "bottom-right corner ray"
-        );
-
-        assert_eq!(
-            intersects_ray_from_center(rect, (Vec2::RIGHT + Vec2::UP).normalized()),
-            pos2(3.0, 1.0),
-            "top-right corner ray"
-        );
-    }
+fn layout_self_edges<'a>(
+    rect: Rect,
+    edges: &'a [&EdgeTemplate],
+) -> impl Iterator<Item = EdgeGeometry> + 'a {
+    edges.iter().enumerate().map(move |(i, edge)| {
+        let offset = (i + 1) as f32;
+        let target_arrow = edge.target_arrow;
+        let anchor = rect.center_top();
+
+        EdgeGeometry {
+            target_arrow,
+            path: PathGeometry::CubicBezier {
+                // TODO(grtlr): We could probably consider the actual node size here.
+                source: anchor + Vec2::LEFT * 4.,
+                target: anchor + Vec2::RIGHT * 4.,
+                // TODO(grtlr): The actual length of that spline should follow the `distance` parameter of the link force.
+                control: [
+                    anchor + Vec2::new(-30. * offset, -40. * offset),
```

Comment:
Does that mean that self edge are _always_ above a node? This feels like a limitation. Ideally, the orientation of the self-edge(s) would also be set by the layout engine (e.g. based on the user provided position if any).

<img width="201" alt="image" src="https://github.com/user-attachments/assets/63278032-5dc3-483b-a8ed-b380e18e08c7">

Probably not for this PR, but maybe worth having an issue and a TODO.

Reply:
Yeah, sorry I should have mentioned. I didn't want to get into edge routing just yet. I've created an issue here to track this: https://github.com/rerun-io/rerun/issues/8291